packages/server/src/utils/gpu-setup.ts

```diff
@@ -0,0 +1,276 @@
+		try {
+			const runtimeCommand = 'docker info --format "{{json .Runtimes}}"';
+			const { stdout: runtimeInfo } = serverId
+				? await execAsyncRemote(serverId, runtimeCommand)
+				: await execAsync(runtimeCommand);
+
+			const runtimes = JSON.parse(runtimeInfo);
+			runtimeInstalled = "nvidia" in runtimes;
+
+			// Check if it's the default runtime
+			const defaultCommand = 'docker info --format "{{.DefaultRuntime}}"';
+			const { stdout: defaultRuntime } = serverId
+				? await execAsyncRemote(serverId, defaultCommand)
+				: await execAsync(defaultCommand);
+
+			runtimeConfigured = defaultRuntime.trim() === "nvidia";
+		} catch (error) {
+			console.debug("Runtime check:", error);
+		}
+
+		// Check Swarm GPU Resources
+		let swarmEnabled = false;
+		let gpuResources = 0;
+
+		try {
+			// Check node resources directly from inspect
+			const nodeCommand =
+				"docker node inspect self --format '{{json .Description.Resources.GenericResources}}'";
+			const { stdout: resources } = serverId
+				? await execAsyncRemote(serverId, nodeCommand)
+				: await execAsync(nodeCommand);
+
+			if (resources && resources !== "null") {
+				const genericResources = JSON.parse(resources);
+				for (const resource of genericResources) {
+					if (
+						resource.DiscreteResourceSpec &&
+						(resource.DiscreteResourceSpec.Kind === "GPU" ||
+							resource.DiscreteResourceSpec.Kind === "gpu")
+					) {
+						gpuResources = resource.DiscreteResourceSpec.Value;
+						swarmEnabled = true;
+						break;
+					}
+				}
+			}
+		} catch (error) {
+			console.debug("Swarm resource check:", error);
+		}
+
+		// Get GPU Model and Memory Info
+		const gpuInfoCommand =
+			"nvidia-smi --query-gpu=gpu_name,memory.total --format=csv,noheader";
+		const { stdout: gpuInfo } = serverId
+			? await execAsyncRemote(serverId, gpuInfoCommand)
+			: await execAsync(gpuInfoCommand);
+
+		const [gpuModel, memoryTotal] = gpuInfo.split(",").map((s) => s.trim());
+
+		// Check CUDA Support
+		const cudaCommand = 'nvidia-smi -q | grep "CUDA Version"';
+		const { stdout: cudaInfo } = serverId
+			? await execAsyncRemote(serverId, cudaCommand)
+			: await execAsync(cudaCommand);
+
+		const cudaMatch = cudaInfo.match(/CUDA Version\s*:\s*([\d\.]+)/);
+		const cudaVersion = cudaMatch ? cudaMatch[1] : undefined;
+		const cudaSupport = !!cudaVersion;
+
+		return {
+			driverInstalled,
+			driverVersion,
+			runtimeInstalled,
+			runtimeConfigured,
+			availableGPUs,
+			swarmEnabled,
+			gpuResources,
+			gpuModel,
+			memoryInfo: memoryTotal,
+			cudaSupport,
+			cudaVersion,
+		};
+	} catch (error) {
+		console.error("Error in checkGPUStatus:", error);
+		return {
+			driverInstalled: false,
+			driverVersion: undefined,
+			runtimeInstalled: false,
+			runtimeConfigured: false,
+			cudaSupport: false,
+			cudaVersion: undefined,
+			gpuModel: undefined,
+			memoryInfo: undefined,
+			availableGPUs: 0,
+			swarmEnabled: false,
+			gpuResources: 0,
+		};
+	}
+}
```

Comment:
I see that this does the job well, however I would like to separate each validation into a separate function, so that it will be easier to find errors in the future.

Example:

The first validation:

```ts
try {
			const driverCommand =
				"nvidia-smi --query-gpu=driver_version --format=csv,noheader";
			const { stdout: nvidiaSmi } = serverId
				? await execAsyncRemote(serverId, driverCommand)
				: await execAsync(driverCommand);

			driverVersion = nvidiaSmi.trim();
			if (driverVersion) {
				driverInstalled = true;
				const countCommand =
					"nvidia-smi --query-gpu=gpu_name --format=csv,noheader | wc -l";
				const { stdout: gpuCount } = serverId
					? await execAsyncRemote(serverId, countCommand)
					: await execAsync(countCommand);

				availableGPUs = Number.parseInt(gpuCount.trim(), 10);
			}
		} catch (error) {
			console.debug("GPU driver check:", error);
		}
```



 can be turn in to this:

```ts
const checkGpuDriver = async (serverId?: string) => {
	let driverVersion: string | undefined;
	let driverInstalled = false;
	let availableGPUs = 0;

	try {
		const driverCommand =
			"nvidia-smi --query-gpu=driver_version --format=csv,noheader";
		const { stdout: nvidiaSmi } = serverId
			? await execAsyncRemote(serverId, driverCommand)
			: await execAsync(driverCommand);

		driverVersion = nvidiaSmi.trim();
		if (driverVersion) {
			driverInstalled = true;
			const countCommand =
				"nvidia-smi --query-gpu=gpu_name --format=csv,noheader | wc -l";
			const { stdout: gpuCount } = serverId
				? await execAsyncRemote(serverId, countCommand)
				: await execAsync(countCommand);

			availableGPUs = Number.parseInt(gpuCount.trim(), 10);
		}
	} catch (error) {
		console.debug("GPU driver check:", error);
	}

	return {
		driverVersion,
		driverInstalled,
		availableGPUs,
	};
};

```
Is more easy to understand and maintainable

Reply:
updated thanks for the review
refactored version:
- Each validation is now a separate function with its own error handling
- The main `checkGPUStatus` function uses Promise.all to run all checks concurrently
- Each function returns only its relevant data
- The final result is composed by spreading all the individual results
- Error handling is maintained at both the individual check level and the overall function level